app/src/main/java/de/westnordost/streetcomplete/data/osm/split/SplitWayUpload.kt

```diff
@@ -0,0 +1,261 @@
+        val newRelationMembers = newWays.map { way ->
+            OsmRelationMember(way.id, originalWayRole, WAY) }.toMutableList()
+        val isOrientedBackwards = originalWay.isOrientedForwardInOrderedRelation(relation, indexOfWayInRelation) == false
+        if (isOrientedBackwards) newRelationMembers.reverse()
+
+        relation.members.removeAt(indexOfWayInRelation)
+        relation.members.addAll(indexOfWayInRelation, newRelationMembers)
+    }
+
+    private fun Node.fetchUpdated() =
+        osmDao.getNode(id) ?: throw ConflictException("Node #$id has been deleted")
+
+    private fun Way.fetchUpdated() =
+        osmDao.getWay(id) ?: throw ConflictException("Way #$id has been deleted")
+
+    private fun Way.fetchParentRelations() = osmDao.getRelationsForWay(id)
+
+    /** returns null if the relation is not ordered, false if oriented backwards, true if oriented forward */
+    private fun Way.isOrientedForwardInOrderedRelation(relation: Relation, indexInRelation: Int): Boolean? {
+        val wayIdBefore = relation.members.findPrevious(indexInRelation) { it.type == WAY }?.ref
+        val wayBefore = wayIdBefore?.let { osmDao.getWay(it) }
+        if (wayBefore != null) {
+            if (isAfterWayInChain(wayBefore)) return true
+            if (isBeforeWayInChain(wayBefore)) return false
+        }
+
+        val wayIdAfter = relation.members.findNext(indexInRelation+1) { it.type == WAY }?.ref
+        val wayAfter = wayIdAfter?.let { osmDao.getWay(it) }
+        if (wayAfter != null) {
+            if (isBeforeWayInChain(wayAfter)) return true
+            if (isAfterWayInChain(wayAfter)) return false
+        }
+
+        return null
+    }
+
+    private fun Relation.fetchViaNodeIds(relationType: String): List<Long>? {
+        val via = findVia(relationType) ?: return null
+        return when (via.type) {
+            WAY -> osmDao.getWay(via.ref)?.nodeIds?.firstAndLast()
+            NODE -> listOf(via.ref)
+            else -> null
+        }
+    }
+}
+
+/** comparator that sorts all the splits from start to end in the way */
+private class SplitWayAtComparator(private val way: Way) : Comparator<SplitWayAtPosition> {
+    override fun compare(split1: SplitWayAtPosition, split2: SplitWayAtPosition): Int {
+        val split1Index = way.nodeIds.indexOf(split1.secondNode.id)
+        val split2Index = way.nodeIds.indexOf(split2.secondNode.id)
+        val diffIndex = split1Index - split2Index
+        if (diffIndex != 0) return diffIndex
+
+        val diffDelta = split1.delta - split2.delta
+        return if (diffDelta < 0) -1 else if (diffDelta > 0) 1 else 0
+    }
+}
+
+private fun createSplitPosition(firstPosition: LatLon, secondPosition: LatLon, delta: Double) =
+    SphericalEarthMath.createTranslated(
+        firstPosition.latitude + delta * (secondPosition.latitude - firstPosition.latitude),
+        firstPosition.longitude + delta * (secondPosition.longitude - firstPosition.longitude))
```

Comment:
It seems to be breaking across 180 line. Likely such objects are already discarded or impossible to appear or it should be rare but I would consider throwing an exception...

`firstPosition.longitude + delta * (secondPosition.longitude - firstPosition.longitude)`

```
firstPosition.longitude = 179
secondPosition.longitude = -179
delta = 0.5
```

Expected: 180 or -180

`179 + 0.5 * (-179 - 179)` = `179 + 0.5 * -179 * 2` = 0`

Reply:
Right, that is a very good point! Thank you!